Feature: Add Infobox to link to individual DC pages

martinctc opened the pull request:
# Summary
This branch *adds an Infobox element to link to the individual DC pages on the website*.

# Changes
The changes made in this PR are:
1. Adds a clickable Infobox.
1. Adds {hkdatasets} to renv.lock - moving data source from Google Sheets to the [package](https://hong-kong-districts-info.github.io/hkdatasets/).
1. Added new Infobox features would phone number, district, and address.
1. Added the ability to search by District, which returns a filtered list of Constituencies to select from.
...

***

# Check
- [x] All checks pass
- [x] That the Infobox link works
- [x] The tutorial works with the new Infobox element.
- [x] The travis.ci and R CMD checks pass.

***

## (OPTIONAL) Note
This "fixes #<issue_number>"

*<other things, such as how to incorporate new changes>*
*<brief summary of the purpose of this pull request>*

martinctc committed:
feat: add infbox to link to individual DC pages

avisionh commented:
Is this ready for me to review, @martinctc?

avisionh commented on inst/app/server.R:
```diff
@@ -122,6 +122,25 @@ server <- function(input, output, session) {
 
+  # InfoBox: Individual DC Page ---------------------------------------------
+  ## Make this a clickable link to the individual DC page
+  output$individual_page <- renderInfoBox(
+    expr = {
+      tags$div(
+        infoBox(value = "",
+                title = "查看更多 / View more",
+                subtitle = "",
+                href = react_data_dropdown()$ind_page,
+                icon = icon(name = "hand-pointer"),
+                color = "green",
+                fill = TRUE,
+                width = 12)
+      ) #div
+    }
+  ) #renderInfoBox
```
Not sure if an `infoBox()` is appropriate for this re-direct because there's a lot of empty space. Maybe make it smaller? 
![image](https://user-images.githubusercontent.com/25527485/90504477-0884bb00-e149-11ea-9607-a289a92f8ce4.png)

martinctc replied:
Hoping to get info like phone numbers and emails this week. Should we sort this out in the same PR, and get the UI sorted out at the same time?

avisionh replied:
Ah, so the plan is to put the phone numbers and emails in this `infoBox()`? If so, makes more sense to add this into the existing PR as they are thematically similar.

avisionh reviewed (changes requested):
Thanks for this, a couple of further thoughts:
1. What do you plan to show on the webpage that's different to the Shiny app?
    + Currently, it's showing the same information as the Shiny app. I think we could build on the extra space the webpage has by including stuff like polling information, voting record etc.
    + Though, should be mixing the website and Shiny app together? If we think about the website and this dashboard as products, then we should ask ourselves, what need does each product serve. I was thinking the website serves the purpose in being a central place to access all our projects whilst the Shiny app serves the purpose of providing District Councillor information.
    + In this regard, would putting the extra information on a separate tab on the Shiny app work better in terms of keeping things focussed? This might work better from a UX perspective as the user can easily go back and change the DC they want more information on.
1. There's a bit of a difference in style between the web page and the Shiny app.
    + This can indicate to the user that the Shiny app is separate to the webpage, when they are actually linked in terms of being within the same umbrella organisation.

avisionh commented on inst/app/ui.R:
```diff
@@ -123,6 +123,9 @@ ui <- dashboardPage(
               infoBoxOutput(outputId = "infobox_constituency", width = NULL)
             ),
+            fluidRow(
+              infoBoxOutput(outputId = "individual_page", width = NULL)
```
Might be useful to add a [shinyhelper](https://github.com/cwthom/shinyhelper) to describe what they can expect to see from being redirected.

martinctc commented:
> Thanks for this, a couple of further thoughts:
> 
> 1. What do you plan to show on the webpage that's different to the Shiny app?
>    
>    * Currently, it's showing the same information as the Shiny app. I think we could build on the extra space the webpage has by including stuff like polling information, voting record etc.
>    * Though, should be mixing the website and Shiny app together? If we think about the website and this dashboard as products, then we should ask ourselves, what need does each product serve. I was thinking the website serves the purpose in being a central place to access all our projects whilst the Shiny app serves the purpose of providing District Councillor information.
>    * In this regard, would putting the extra information on a separate tab on the Shiny app work better in terms of keeping things focussed? This might work better from a UX perspective as the user can easily go back and change the DC they want more information on.
> 2. There's a bit of a difference in style between the web page and the Shiny app.
>    
>    * This can indicate to the user that the Shiny app is separate to the webpage, when they are actually linked in terms of being within the same umbrella organisation.

All great points! If I'm completely honest, this was driven by a desire to see whether it's possible to generate programmatically an individual page for each District Councillor, and the ability to link to each individual DC (currently not possible with the Shiny app). But if you think about it, it could work this way (elaborating on your suggestion): 

(1) The app acts as a starting point for users to get basic information about the DC, which is pretty much what we've got now. If they want to see more, they can visit the individual DC page which is linked on the app. The app is supposed to make it _easier_ for the user to search for DCs, versus searching on the site; e.g. it's not possible to do checkbox toggles on the website. 

(2) The individual page, as you said, could provide much richer detail about the DC, including charts, voting records, or even a line of introduction that we could ask DCs to submit. It could well be a mini portfolio for the DC - the possibilities are quite endless. 

I agree also with the styling point, but that's more minor and we can sort out down the line - at least we've got the same logo being shown!

avisionh commented:
> > Thanks for this, a couple of further thoughts:
> > 
> > 1. What do you plan to show on the webpage that's different to the Shiny app?
> >    
> >    * Currently, it's showing the same information as the Shiny app. I think we could build on the extra space the webpage has by including stuff like polling information, voting record etc.
> >    * Though, should be mixing the website and Shiny app together? If we think about the website and this dashboard as products, then we should ask ourselves, what need does each product serve. I was thinking the website serves the purpose in being a central place to access all our projects whilst the Shiny app serves the purpose of providing District Councillor information.
> >    * In this regard, would putting the extra information on a separate tab on the Shiny app work better in terms of keeping things focussed? This might work better from a UX perspective as the user can easily go back and change the DC they want more information on.
> > 2. There's a bit of a difference in style between the web page and the Shiny app.
> >    
> >    * This can indicate to the user that the Shiny app is separate to the webpage, when they are actually linked in terms of being within the same umbrella organisation.
> 
> All great points! If I'm completely honest, this was driven by a desire to see whether it's possible to generate programmatically an individual page for each District Councillor, and the ability to link to each individual DC (currently not possible with the Shiny app). But if you think about it, it could work this way (elaborating on your suggestion):
> 
> (1) The app acts as a starting point for users to get basic information about the DC, which is pretty much what we've got now. If they want to see more, they can visit the individual DC page which is linked on the app. The app is supposed to make it _easier_ for the user to search for DCs, versus searching on the site; e.g. it's not possible to do checkbox toggles on the website.
> 
> (2) The individual page, as you said, could provide much richer detail about the DC, including charts, voting records, or even a line of introduction that we could ask DCs to submit. It could well be a mini portfolio for the DC - the possibilities are quite endless.
> 
> I agree also with the styling point, but that's more minor and we can sort out down the line - at least we've got the same logo being shown!

I like the use of the website because of the different webpages and so if we incorporate GA tracking on it also, then we really can see how people visit different pages and then further down the line, if we can get this information from BigQuery, we could do some user-journeys analysis. Currently, the Shiny app exists as 'one' page, so we can't actually see what people do on the app.

However, my questions surrounding the use of the website in this way are:
- From the perspective of the website as a product, what is the focus of it? My understanding was it acts as a gateway for all our projects, providing high-level descriptions for them and links to them. (Like the diagram below) Should we start mixing it with DC information, then we will need to revise what the website's aim is.
    + Thus, the website is like the public frontend to the GitHub organisation that you set-up.
- The ease of switching back and forth between the Shiny app and the website so users can look at the more granular detail for each DC.

![image](https://user-images.githubusercontent.com/25527485/90540419-1baf7f00-e179-11ea-8070-82f7812f660e.png)

martinctc commented:
> * From the perspective of the website as a product, what is the focus of it? My understanding was it acts as a gateway for all our projects, providing high-level descriptions for them and links to them. (Like the diagram below) Should we start mixing it with DC information, then we will need to revise what the website's aim is.
>   
>   * Thus, the website is like the public frontend to the GitHub organisation that you set-up.
> * The ease of switching back and forth between the Shiny app and the website so users can look at the more granular detail for each DC.
> 
> ![image](https://user-images.githubusercontent.com/25527485/90540419-1baf7f00-e179-11ea-8070-82f7812f660e.png)

Just on this, I think you're right that the website is intended to be a gateway for all our projects. But since our projects are unified by a common theme (i.e. Hong Kong districts), having individual pages for the District Councillors is actually fine because we have the flexibility to connect traffic collision insights in the individual DC pages (for example). So it need not conflict with the original intent of the website? I'm kind of inclined to let these individual pages evolve more organically and see what we can populate on these - simply being able to have everything in one place by constituency is adding so much value already.

gabtam55 commented:
Just went through your conservations. Am I correct that we are thinking to add the new data (e.g. voting records, polling information) to the website instead of the shiny app?

I personally prefer these new data to be added to the shiny app.
- Like @avisionh said, the website is a gateway to all our apps. We may want to keep it at high level to accomplish only simple tasks such as providing district councillors' contact information quickly, releasing blog articles, providing a 'menu' including all our apps (when we have a lot of them in the future :p). Currently I see each app to provide information on a specific topic of a district council (e.g. district councillors analysis, traffic analysis). They are reserved for people (e.g. traffic experts, district councillors, journalists) who are interested in a particular topic and would like more in-depth data visualisation.
- Once we have more data related to each district councillor, the interactive features on shiny will become handy for users to draw insights from the data (which will more difficult to achieve on the website?). Example user stories could be: 1) Finding out district councillors who won the election by small amount of votes; 2) Finding out whether district councillors from a particular political party tend to have poorer attendance to their district council regular meeting.

I do acknowledge the limitation that we cannot do much user-tracking with shiny app.

martinctc committed:
feat: add hkdatasets to renv.lock

martinctc committed:
fix: fix package issue with renv.lock

martinctc committed:
feat: migrate from googlesheets4 set up to hkdatasets

martinctc committed:
feat: add phone number InfoBox

martinctc committed:
feat: add address, district

martinctc committed:
feat: allow nested select

Changed architecture such that the Constituency selection is done via ui.R

martinctc committed:
fix: fixed bug with filter

martinctc committed:
fix: add default selection

Issue with filter warning still exists

avisionh commented on inst/app/rsconnect/shinyapps.io/hkdistricts-info/dashboard-hkdistrictcouncillors-dev.dcf:
```diff
@@ -0,0 +1,12 @@
+url: https://hkdistricts-info.shinyapps.io/dashboard-hkdistrictcouncillors-dev/
+when: 1598030600.53705
+asMultiple: FALSE
+asStatic: FALSE
```
Can you delete this file please? It's an artefact of the deployment to shinyapps.io so not strictly necessary for the repo.

martinctc replied:
Would you agree to add this to the .gitignore? It does keep generating itself when I deploy the app

avisionh replied:
Yes, that's a better good idea! Delete it, add to `gitignore` and then commit it.

avisionh commented on inst/app/server.R:
```diff
@@ -33,10 +33,12 @@ server <- function(input, output, session) {
                    label = "請選擇或輸入選區 / Please type or select a constituency",
-                   choices = react_region_dropdown()$DropDownText)
+                   choices = react_region_dropdown()$DropDownText,
+                   selected = "A01: 中環 / CHUNG WAN")
```
This is nice, more explicit.

avisionh commented on inst/app/global.R:
```diff
@@ -11,8 +11,8 @@
 ## data wrangling
 library(dplyr)
 library(tidyr)
-library(googlesheets4)
```
As we're removing `googlesheets4` from the repo, can you also do likewise for the `renv.lock` and `DESCRIPTION` files?

martinctc replied:
@avisionh What's your view on keeping the google sheets clean up code in the repo? Sure, it's no longer used, but it played a part in the development of the app? We could have some sort of legacy / archive subfolder?

avisionh replied:
I'd say it is better to go into the [hkdatasets](https://github.com/Hong-Kong-Districts-Info/hkdatasets) package.

martinctc replied:
Agreed. Have moved this to [hkdatasets](https://github.com/Hong-Kong-Districts-Info/hkdatasets).

avisionh reviewed (changes requested):
Thanks for the good work! Couple of extra things to consider!

martinctc committed:
chore: remove googlesheets4 from DESCRIPTION and renv.lock

martinctc committed:
feat: add spinner to constituency_dropdown rendering

martinctc committed:
chore: add shinyapps deploy log to .gitignore

martinctc committed:
chore: suppress warnings with tags$style

Added some code documentation on input_region

avisionh commented on inst/app/rsconnect/shinyapps.io/hkdistricts-info/dashboard-hkdistrictcouncillors-dev.dcf:
```diff
@@ -0,0 +1,12 @@
+url: https://hkdistricts-info.shinyapps.io/dashboard-hkdistrictcouncillors-dev/
+when: 1598113019.08334
+asMultiple: FALSE
+asStatic: FALSE
```
Can you also delete this and then commit the deletion, please?

avisionh commented on inst/app/global.R:
```diff
@@ -11,8 +11,8 @@
 library(tidyr)
-library(googlesheets4)
 library(sf)
+library(hkdatasets) # Installed with tarball. See https://github.com/Hong-Kong-Districts-Info/hkdatasets
```
No need for the comment because it's automatically downloaded from the `renv.lock` file.
```suggestion
library(hkdatasets)
```

avisionh commented on inst/app/server.R:
```diff
@@ -16,12 +16,42 @@ server <- function(input, output, session) {
+        } else {
+          
+          
+          data_select <- dplyr::filter(.data = data_master_raw,
+                                       District == input$input_region)
```
```suggestion
          data_select <- filter(.data = data_master_raw,
                                             District == input$input_region)
```

avisionh commented on inst/app/server.R:
```diff
@@ -16,12 +16,42 @@ server <- function(input, output, session) {
+  # renderUI for second dropdown for constituency
+  # dropdown filters according to selected region
+  # method implemented in https://stackoverflow.com/questions/45975959/create-reactive-selectinput-flexdashboard-with-shiny
+  output$constituency_dropdown <- renderUI({
+    selectizeInput(inputId = "input_dropdowntext",
+                   label = "請選擇或輸入選區 / Please type or select a constituency",
+                   choices = react_region_dropdown()$DropDownText,
+                   selected = "A01: 中環 / CHUNG WAN")
```
The warnings are quite annoying, especially when they show up in the app. I've had a go and overcame this. You can resolve the warnings via inputting `req(input$input_region)` in the reactive above.
https://shiny.rstudio.com/articles/req.html

avisionh replied:
Also, could this `renderUI` part go into the `# Tab: Overview of a DC` section? It's currently misplaced in the `# ----- REACTIVES ----- #` section.

avisionh replied:
Own example here:
https://github.com/Hong-Kong-Districts-Info/dashboard-hkdistrictcouncillors/pull/70/commits/331cf99e2122be9806204e68e1dbc38ff04b34dc

martinctc replied:
Great fix! This has worked. Also moved the `renderUI` to the appropriate section.

avisionh commented on inst/app/server.R:
```diff
@@ -16,12 +16,42 @@ server <- function(input, output, session) {
-                            DropDownText == input$input_dropdowntext)
       
+      # data_master_raw[data_master_raw$DropDownText %in% input$input_dropdowntext,]
+      data_select <- dplyr::filter(.data = data_master_raw,
+                                   DropDownText == input$input_dropdowntext)
```
```suggestion
      data_select <- filter(.data = data_master_raw,
                                        DropDownText == input$input_dropdowntext)
```

avisionh reviewed (changes requested):
I can really see what you mean about the warnings, super annoying! I provided the link to an example that might get round this problem. Hopefully it's relevant.

avisionh commented on inst/app/ui.R:
```diff
@@ -89,6 +89,12 @@ ui <- dashboardPage(
       tags$head(includeHTML(("google-analytics.html")))
     ),
     
+    # Suppress Warnings
+    tags$style(type="text/css",
+               ".shiny-output-error { visibility: hidden; }",
+               ".shiny-output-error:before { visibility: hidden; }"
+    ),
+    
```
No longer need to suppress warnings if we use `req(input$<name>`)
```suggestion

```

martinctc committed:
fix: remove deployment file for shinyapps.io

martinctc committed:
docs: remove unnecessary comment in global.R

Already covered in `renv.lock`

Co-authored-by: A Ho <a_vision@hotmail.co.uk>

martinctc committed:
fix: tidy code alignment in server.R

Co-authored-by: A Ho <a_vision@hotmail.co.uk>

martinctc committed:
fix: restructure server.R and remove warnings

- Moved output$constituency_dropdown to the correct part of the script
- added `req()` to remove warning messages

martinctc committed:
docs: tidy up code on server.R

martinctc committed:
fix: remove warning suppression css on ui.R

martinctc committed:
chore: removed Google Sheets script

These have now been moved to the hkdatasets repo

martinctc committed:
docs: add hkdatasets reference on README.md

martinctc committed:
feat: optimise for responsive

avisionh committed:
fix: Remove specification of Linux distributions

This is to get CI-CD working. Seems to be a related issue with the bionic distribution being broken and speculating that trusty is the same.

avisionh commented:
Great work on making the iframe more responsive - happy with this!

Just to note, since we have moved the GSheets cleaning, then we want the unit-testing associated with this to also go into the [hkdatasets](https://github.com/Hong-Kong-Districts-Info/hkdatasets) package.  I've done this [here](https://github.com/Hong-Kong-Districts-Info/hkdatasets/pull/2)
